inc/ctest.h

```diff
@@ -272,39 +229,17 @@ static void MU_C2(toStringType,_ToString)(char* string, size_t bufferSize, cType
 void do_jump(jmp_buf *exceptionJump, const volatile void* expected, const volatile void* actual);
 
 #define CTEST_ASSERT_ARE_EQUAL(type, A, B, ...) \
-do { \
-    const type A_value = (const type)(A); \
```

Comment:
>st type A_v [](start = 7, length = 11)

I understand the problem is the temporary generation. I have the 2014

Temporary objects are destroyed as
the last step in evaluating the full-expression (6.9.1) that (lexically) contains the point where they were
created.

This is from the 2020 standard but 2014 has the same statement.

So, can we have an expression built that uses the ternary operator?
The reason I am suggesting this is to avoid needing to update all the usages to add CTEST_DECLARE_EQUALITY_ASSERTION_FUNCTIONS_FOR_TYPE and the equivalent DEFINE macros.

I am pasting here a simple program that might help exemplify it:

#include <string>

static int times_called;

int func1(const char* a)
{
    (void)printf(a);
    (void)printf("\r\n");
    times_called++;
    return 0;
}

int func2(const char* a, const char* b)
{
    if (strcmp(a, b) == 0)
    {
        printf("same\r\n");
        return 1;
    }
    else
    {
        printf("not same\r\n");
        return 0;
    }
}

#define TEST_MACRO(type, a, b) \
do \
{ \
    static type a_value; \
    static type b_value; \
    func1(a_value = (type)a) ?  \
        func1(b_value = (type)b) ?  \
            func2(a_value, b_value) : func2(a_value, b_value) : \
        func1(b_value = (type)b) ? func2(a, b) : func2(a, b); \
} while (0)

typedef char* ascii_ptr;

int main(void)
{
    std::string x = "a";
    std::string y = "b";

    TEST_MACRO(ascii_ptr, x.c_str(), y.c_str());

    printf("Times called for func1 = %d", times_called);

    return 0;
}

Reply:
Disclaimer: I have not tested this in the context of ctest. I hope that my assumption that the ternary operator will make the full expression enclose the calls to _ToString and _Compare, which means that the strings used would be valid since the temporaries would not get destroyed too early.

---
In reply to: [637702898](https://github.com/Azure/ctest/pull/91#discussion_r637702898) [](ancestors = 637702898)

Reply:
Thanks for the suggestion. I'll look into it tomorrow.

Reply:
We can almost get there by using a combination of ternary and comma operator, as shown below:

```
#define CTEST_ASSERT_ARE_EQUAL(type, A, B, ...) \
do { \
    type A_value; \
    type B_value; \
    char expectedString[1024]; \
    char actualString[1024]; \
    char* ctest_message = GET_MESSAGE(__VA_ARGS__); \
    MU_C2(type,_Compare)(A_value = (A), B_value = (B)) ? \
        /* Sequence when A is not equal to B*/ \
        (MU_C2(type,_ToString)(expectedString, sizeof(expectedString), A_value), \
         MU_C2(type,_ToString)(expectedString, sizeof(expectedString), B_value), \
         (void)printf("  Assert failed in line %d %s Expected: %s, Actual: %s\n", __LINE__, (ctest_message == NULL) ? "" : ctest_message, expectedString, actualString), \
         ctest_sprintf_free(ctest_message), \
         g_CurrentTestFunction != NULL ? (*g_CurrentTestFunction->TestResult = TEST_FAILED, (void)0) : (void)0, \
         do_jump(&g_ExceptionJump, expectedString, actualString)) : \
        /* Sequence when A is equal to B (do nothing)*/ \
        ctest_sprintf_free(ctest_message); \
} while (0)
```

However, this method forces the use of default-constructed `A_value` and `B_value`. I don't recommend this because it breaks the CTEST API with types that don't have a default constructor, reference types (because references must always be initialized), and to some degree with `const` values. Instead, I would recommend keeping the `AssertAreEqual` and `AssertAreNotEqual` functions, but automatically generating their definitions in `CTEST_COMPARE`. This approach keeps the API flexible and addresses you concern about  the need to modify all usages. I've pushed new code just now that does this.

Thoughts?

Reply:
I see. Ok, I appreciate you trying it. I don't have better ideas at the moment to avoid the issues for reference types types without default constructor.
Not sure which const cases break but maybe it does not even matter.

What would be needed though is adding cases in the cpp test suite for each of the 3:
- types with default constructor
- reference type
- one of the const cases you know won't work

The reason is that if I or somebody in the team gets to tinker with this and imrpove it, we would not want to break those cases again.

---
In reply to: [638396396](https://github.com/Azure/ctest/pull/91#discussion_r638396396) [](ancestors = 638396396)

Reply:
The constness problem was related to `void_ptr` not having `const`. Sometimes if using CTEST_ASSERT_ARE_EQUAL with void_ptr, there would be a warning emitted when passing in a const ptr because the constness would differ when assigning the const ptr to the non const type. I fixed this and added a test for this as well.

Reply:
Perfect, thanks!

---
In reply to: [639391277](https://github.com/Azure/ctest/pull/91#discussion_r639391277) [](ancestors = 639391277)